src/app/qgisapp.cpp

```diff
@@ -6452,8 +6452,12 @@ void QgisApp::refreshFeatureActions()
   QList<QgsAction> actions = vlayer->actions()->actions( QStringLiteral( "Canvas" ) );
   Q_FOREACH ( const QgsAction &action, actions )
   {
+    if( vlayer->readOnly() && action.isEnabledOnlyWhenEditable() )
```

Comment:
I think this should be vlayer->isEditable(), not readOnly() - e.g. hide the actions if not editable as opposed to showing them as disabled. Best to keep the ui as clutter-free as possible when we can.

Reply:
My thought was: Hide them, when they never should appear. Disable them, when they would be available in editable mode. So the user sees, that there would be something, when he toggles edit mode (similar to the Tools like "Add Feature" etc.).
But I could change it to not even list them as well...

Reply:
My personal preference would be to hide them. I really don't want to see actions become a huge mash of inbuilt features, since I think the primary use case for them is user-created actions, and I want to ensure that those user created actions remain prominent and aren't lost amongst many other inbuilt actions. 

@nirvn what do you think?

Reply:
Yep agreed. I was about to write this very comment once at office.

Reply:
@nyalldawson - sorry, but I disagree here. From the QGIS manual "QGIS provides the ability to perform an action based on the attributes of a feature. This can be used to perform * any number * of actions, for example, running a program with arguments built from the attributes of a feature or passing parameters to a web reporting tool." (https://docs.qgis.org/2.18/en/docs/user_manual/working_with_vector/vector_properties.html#actions-properties) - clearly this does not exclude internal QGIS functionality - if - it is based on a feature attribute. Which clearly is the case here. Based on a feature, we create a new feature, that has the same attributes. I don't think it is clearly defined that Actions are only for user-created actions, and not for builtin functionality. Plus - the Atlas features is also a builtin action that is always - just there.

The primary reason why Matthias and I suggested the "duplicate feature" functionality to be an action is the fact, that actions are - as far as I know - the only easy way to have right-click context menu functionality on table rows.
Plus - if something is defined as an action it appears at many places automatically (as buttons, in form menu, right-click on table rows, in action tool, identify tool, etc. without having to program a lot.

Reply:
ok - I am probably wrong that actions are the only way to have a context menu. I see that other - non-action based entries are here.
But still it is convenient, that something defined as an Action appears at many places where one wants to do something based on a feature or based on a features attribute.

Reply:
Yes, but you're coming from the viewpoint of someone who will use that feature heavily. For many others it's just extra interface noise cluttering the actions they've defined themselves. That's why I think we hide when they can't be used, because at least that helps mitigate the impact. 

I feel the same with the atlas now, but it was added when actions were only exposed in about half the places they are now. And the atlas functionality only shows in very narrow circumstances - you have to have a layout with an atlas defined on that layer. Thats in contrast to these two actions which are always present for every vector layer.

Reply:
Interface cluttering vs. availability of functionality is always a tricky thing to get right (and probably impossible to completely get right due to the wide range of users QGIS has)

In this case here I feel, that hiding something based on layer editable state is only masquerading an issue and will not help. I personally often remember menu actions based on their position (e.g. 3rd from the bottom) and hiding and showing them based on the editable state breaks this workflow.

I am in favor of the approach proposed in here and if the concerns about too many actions continue, let's find a better solution like configuration options for enabling and disabling or nested menus.